### PR TITLE
Updates to ABCI_Proof PR

### PR DESCRIPTION
**Pull request by jaekwon**

### CODE REVIEW

File: cmd/abci-cli/abci-cli.go [hunk 1]
```diff
@@ -228,9 +205,10 @@ func cmdEcho(c *cli.Context) error {
 	if len(args) != 1 {
 		return errors.New("Command echo takes 1 argument")
 	}
-	res := client.EchoSync(args[0])
-	rsp := newResponse(res, string(res.Data), false)
-	printResponse(c, rsp)
+	resEcho := client.EchoSync(args[0])
+	printResponse(c, response{
+		Data: resEcho.Data,
```
Comment (melekes) [inline cmd/abci-cli/abci-cli.go]: why you wrap `resInfo.Data` into `[]byte` down below and do not wrap `resEcho.Data` here? small inconsistency caught my eye

**Comment by ethanfrey** [comment]
Hey Jae, in general, I agree with the changes.  It makes both queries and proofs much richer structures. And seems much more extensible as a general interface.

However, I have some questions, which may just be resolved with a bit more documentation.

1. **QueryRequest** what are the semantics of both `Path` and `Query` and how do they play together?  Guessing from a discussion we had, `Path` informs the application how to interpret the `Query` bytes.  `ByKey`, `ByName`, etc... I assume these are app-specific?  But it would be nice to have one reserved keyword for `ByKey`, more info in part 2. A bit more explanation would be very helpful, so developers of abci apps know how to properly organize the apps.

2. **QueryResponse** Nice to add `Height` even for queries.  But what are the semantics of the `Proof` and `Data` Bytes?  If I make a query with a path other than `ByKey`, then does `Data` contain both the key and the value?  Or just the value? How are they packed together?? 
  * In the `Proof` case, one can only provide proofs based on queries for the primary key.  I can provide a proof for the state or absence of one key in the tree without problems.  However, it is impossible to provide a proof that a given name is not in the tree (or is only in there the two times I return, not a third).  Thus, one could only safely set the `Proof` flag on the query if the `Path` is `ByKey`? Or should one return proofs for the 1 or more results?
  * Also, what is the proof now?  It contains a list of nodes/hashes from the AppHash (root) down to the specified key/value pair.  That is clear.  However, we then need to hash the key/value pair to verify it matches the hash for the last node in this path.  Do we embed the key/value pair in the proof as `IAVLProof` does now? Or do we abstract the proof to just the path, and then return Key and Value in the query, leaving it to the app to hash them together and verify they match the proof?  This seems the best way to avoid redundant data, but involves changing a bit of the data structures and logic (not too bad).

Here is my proposal to address these points, please take what pieces seem relevant to you into this PR:

#### Query
  * __Arguments__:
    * `Query ([]byte)`: The query request bytes
    * `Path (string)`: Path of request
      * This is like the prefix bytes to the query and determines how to interpret it
      * All abci apps *must* interpret `/` as a query by key on the store
      * Apps *may* allow queries over specific types like `/accounts/` or `/votes/`
      * Apps *may* allow queries over fields of the data like `/accounts/?name=` to search all accounts with a name that matches the query bytes
    * `Height (uint64)`: The block height for which you want the query (default=0 returns data for the latest committed block)
    * `Prove (bool)`: Return Merkle proof with response, returns error if the path is not by key (eg. `/accounts/?name=`)
  * __Returns__:
    * `Code (uint32)`: Response code		
    * `Key ([]byte)`: The key of the matching data (useful for proofs especially)
    * `Value ([]byte)`: The value of the matching data
    * `Log (string)`: Debug or error message
    * `Height (uint64)`: The block height from which data was derived
    * `Proof ([]byte)`: Proof for the data, if requested.  This is a list of node hashes from the root (AppHash) down to the specified node.  The last element is hash(key+value), which can be used to verify the data in `Key` and `Value` is correct.	  

This adds a lot more semantics that `[]byte` to the return value and we must make it clear to developers what information should be returned where, so clients can be written with the expectation of a consistent interface between one app and another. 

The one question I have this this is how it would handle returning multiple values (if I search by name for example).  It doesn't really fit in this schema.  Or the last one either.

**Comment by ethanfrey** [comment]
One more thought, if we are providing a structure to the queries, so we can have a general-purpose light client that doesn't just treat this as unknowable opaque bytes, maybe we should provide a defined structure for the proof bytes. Maybe with a protobuf encoding, and algorithm to verify. Currently, one must call into `go-merkle/IAVLProof` to validate the proof, which is not the best for inter-language inter-operability.

Ex. spec proof is an array of `nodehash`, followed by the roothash.  Each `nodehash` contains a 20-byte hash, 1 byte to flag left or right, and 3 bytes of buffer (so we get a nice 8-byte alignment).  The elements of the proof start closest to the data, working "up" towards the root. We can specify an validation algorithm (in pseudocode/python) like:

```
// start with the data hash
hash = ripemd160(concat(key, value))
// go up each step and combine the hashes of the lower level and hash them together
for nodehash in proof:
  if nodehash.left:
    tohash = concat(hash, nodehash.hash)
  else:
    tohash = concat(nodehash.hash, hash)
  hash = ripemd160(tohash)
// make sure the final result is what the proof said
valid = (hash == roothash)
// TODO: make sure this roothash is the apphash in a validated block....
```

The data hash is ``.

This is basically what is in IAVLProof, but we could make it more general, so other merkle implementations in other languages are supported.  Although we do hardcode the hash function to ripemd160.

File: cmd/abci-cli/abci-cli.go (hunk 1 above)
Comment (jaekwon) [inline cmd/abci-cli/abci-cli.go]: resInfo.Data is a string.  Arguably resInfo.Data should be a []bytes or be renamed something to be consistent, but maybe someone should create a new pull request with proposed changes.  I'm not sure what's best.

**Comment by jaekwon** [comment]
Here's what I have so far in my local dev branch:

```
#### Query
  * __Arguments__:
    * `Data ([]byte)`: Raw query bytes.  Can be used with or in lieu of Path.
    * `Path (string)`: Path of request, like an HTTP GET path.  Can be used with or in liue of Data.
      * Apps MUST interpret '/store' as a query by key on the underlying store.  The key SHOULD be specified in the Data field.
      * Apps SHOULD allow queries over specific types like '/accounts/...' or '/votes/...'
    * `Height (uint64)`: The block height for which you want the query (default=0 returns data for the latest committed block)
    * `Prove (bool)`: Return Merkle proof with response if possible
  * __Returns__:
    * `Code (uint32)`: Response code
    * `Value ([]byte)`: The value of the matching data
    * `Log (string)`: Debug or error message
    * `Height (uint64)`: The block height from which data was derived
    * `Proof ([]byte)`: Proof for the data, if requested.
  *Please note* The current implementation of go-merkle doesn't support querying proofs from past blocks, so for the present moment, any height other than 0 will return an error.  Hopefully this will be improved soon(ish)
```

**Pull request merged**